mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp

```diff
@@ -192,15 +202,30 @@ struct ToQuantumComputationPattern final : mlir::OpRewritePattern<AllocOp> {
       throw; // Rethrow the exception if it's not the expected one.
     }
     // Update `currentQubitVariables` with the new qubit values.
-    for (size_t i = 0; i < ctrlInsIndices.size(); i++) {
-      currentQubitVariables[ctrlInsIndices[i]] = outs[i + 1];
+    for (size_t i = 0; i < posCtrlInsIndices.size(); i++) {
+      currentQubitVariables[posCtrlInsIndices[i]] = outs[i + 1];
+    }
+    for (size_t i = 0; i < negCtrlInsIndices.size(); i++) {
+      currentQubitVariables[negCtrlInsIndices[i]] =
+          outs[i + 1 + posCtrlInsIndices.size()];
     }
     currentQubitVariables[targetIndex] = outs[0];
 
     // Add the operation to the QuantumComputation.
-    circuit.emplace_back<qc::StandardOperation>(
-        qc::Controls{ctrlInsIndices.cbegin(), ctrlInsIndices.cend()},
-        targetIndex, opType);
+    qc::Controls controls;
+    std::transform(posCtrlInsIndices.cbegin(), posCtrlInsIndices.cend(),
+                   std::inserter(controls, controls.end()),
+                   [](const size_t index) {
+                     return qc::Control{static_cast<qc::Qubit>(index),
+                                        qc::Control::Type::Pos};
+                   });
+    std::transform(negCtrlInsIndices.cbegin(), negCtrlInsIndices.cend(),
+                   std::inserter(controls, controls.end()),
+                   [](const size_t index) {
+                     return qc::Control{static_cast<qc::Qubit>(index),
+                                        qc::Control::Type::Neg};
+                   });
```

Comment:
I can't really help myself, but I personally think this is expressed much clearer as a simple range-based for loop, that emplaces the respective controls.
Something like
```c++
for (const auto& index: posCtrlInsIndices) {
  controls.emplace(static_cast<qc::Qubit>(index), qc::Control::Type::Pos);
}
for (const auto& index: negCtrlInsIndices) {
  controls.emplace(static_cast<qc::Qubit>(index), qc::Control::Type::Neg);
}
```